crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -52,13 +53,35 @@ impl Solution {
         score: SolverScore,
         weth: eth::WethAddress,
     ) -> Result<Self, SolutionError> {
+        let score_cip38 = SolverScoreCIP38 {
+            surplus: {
+                let mut surplus = HashMap::new();
+                for trade in trades.iter() {
+                    match trade.surplus(&prices, weth) {
+                        Ok(eth::Asset { token, amount }) => {
+                            surplus
+                                .entry(token)
+                                .or_insert(eth::TokenAmount(eth::U256::zero()))
+                                .0 += amount.0;
+                        }
+                        Err(_err) => {
+                            // todo CIP38 enable
+                            // return Err(SolutionError::Scoring(err))
+                        }
+                    }
+                }
+                surplus
+            },
+        };
```

Comment:
> Why do we need to create the SolverScoreCIP38 struct here?

It needs to be called before the `solution.with_protocol_fees()` is called, because we want our score to be based on the surplus before protocol fees are taken.

> And what benefit do we have from storing the list of surpluses rather than normalizing it directly into a single native token value

The function is also called for `/quote` where external prices vector is not available. See PR description ☝️

Reply:
Added comment explaining this.

Reply:
> It needs to be created before the solution.with_protocol_fees() is called, because we want our score to be based on the surplus before protocol fees are taken.

🙈 In my opinion these kind of implicit timing assumptions make the code extremely hard to reason about and hacky. In this case we should maybe introduce a new type SettlementWithFees to make it clear that some operations aren't supported any longer?

> The function is also called for /quote where external prices vector is not available. See PR description ☝️

Honestly, I'm really unhappy on how this PR changes the driver codebase which we have spent so much ressources on to get to where we are. We now have two score fields ("old score" is a really bad name imo) implicit conditional logic for quote vs solve codepaths (we secretly expect the exact same auction that we use when calling `Solution:new` to have native eth prices when used in `to_native_score`) and are introducing assumptions on the order on how things are called. This is too hacky and reminds me of the convoluted logic in the solver crate which we are trying to kill.

If computing the score before we call `with_protocol_fees` is really necessary (ie we cannot compute it based on the fee information we have available, which is something that has to happen at some point anyways) then I would suggest to extend the existing `SolverScore` to also have a `Surplus` variant and use some field on the auction to decide in `domain::Solution` whether to use the solver provided score or compute our own.

As for the quote path, I think we should either invent a native price or be fine with not computing a score in case it is absent.

Reply:
Moved PR to on hold. Will try to calculate the surplus based on the final solution with applied protocol fees which should remove the duplicate score and most of the confusion.

Reply:
Is this how you preferred to calculate the score: https://github.com/cowprotocol/services/pull/2421/files#diff-4c8f610bdbea68810a9395f71782c8a9fb5a6931e9e814a2f5f9b33bdc847e2eR101-R128

Reply:
Another thing I could try is to encode the `Settlement` into `EncodedSettlement` and have a completely new protocol fee and surplus calculation for it, which should guarantee that the driver and autopilot end up with identical values?